js/compiler_tests.ts

```diff
@@ -0,0 +1,313 @@
+  }
+};
+
+const tsMock: compiler.Ts = {
+  createLanguageService(host: ts.LanguageServiceHost): ts.LanguageService {
+    return {} as ts.LanguageService;
+  },
+  formatDiagnosticsWithColorAndContext(
+    diagnostics: ReadonlyArray<ts.Diagnostic>,
+    host: ts.FormatDiagnosticsHost
+  ): string {
+    return "";
+  }
+};
+const serviceMock = {
+  getCompilerOptionsDiagnostics(): ts.Diagnostic[] {
+    return [];
+  },
+  getEmitOutput(fileName: string): ts.EmitOutput {
+    console.log("getEmitOutput", fileName);
+    if (fileName in emittedFiles) {
+      return {
+        outputFiles: [{ text: emittedFiles[fileName] }] as any,
+        emitSkipped: false
+      };
+    }
+    return { outputFiles: [], emitSkipped: false };
+  },
+  getSemanticDiagnostics(fileName: string): ts.Diagnostic[] {
+    return [];
+  },
+  getSyntacticDiagnostics(fileName: string): ts.Diagnostic[] {
+    return [];
+  }
+};
+const windowMock = {
+  define: undefined
+};
+const mocks = {
+  _globalEval: globalEvalMock,
+  _log: logMock,
+  _os: osMock,
+  _ts: tsMock,
+  _service: serviceMock,
+  _window: windowMock
+};
+
+const originals: { [prop: string]: any } = {};
+
+let compilerInstance: compiler.DenoCompiler;
+
+test(function compilerExists() {
+  assert(DenoCompiler != null);
+  assert(DenoCompiler.instance() != null);
+});
+
+test(function compilerTestsSetup() {
+  compilerInstance = DenoCompiler.instance();
+  assert("_globalEval" in compilerInstance);
+  assert("_log" in compilerInstance);
+  assert("_os" in compilerInstance);
+  assert("_ts" in compilerInstance);
+  assert("_service" in compilerInstance);
+  assert("_window" in compilerInstance);
+  originals._globalEval = (compilerInstance as any)._globalEval;
+  originals._log = (compilerInstance as any)._log;
+  originals._os = (compilerInstance as any)._os;
+  originals._ts = (compilerInstance as any)._ts;
+  originals._service = (compilerInstance as any)._service;
+  originals._window = (compilerInstance as any)._window;
+  Object.assign(compilerInstance, mocks);
+});
+
+test(function compilerMakeDefine() {
+  const moduleMetaData = new ModuleMetaData(
+    "/root/project/foo/bar.ts",
+    fooBarTsSource,
+    fooBarTsOutput
+  );
+  const localDefine = compilerInstance.makeDefine(moduleMetaData);
+  let factoryCalled = false;
+  localDefine(
+    ["require", "exports", "compiler"],
+    (_require, _exports, _compiler): void => {
+      factoryCalled = true;
+      assert(
+        typeof _require === "function",
+        "localRequire should be a function"
+      );
+      assert(_exports != null);
+      assert(
+        Object.keys(_exports).length === 0,
+        "exports should have no properties"
+      );
+      assert(compiler === _compiler, "compiler should be passed to factory");
+    }
+  );
+  assert(factoryCalled, "Factory expected to be called");
+});
+
+test(function compilerRun() {
+  const result = compilerInstance.run("foo/bar.ts", "/root/project");
+  assert(result != null);
+});
```

Comment:
Would be good to check how many times codeFetch and codeCache are called here.

Reply:
👍 Yeah, this was just a basic "hack" at this point... going to introspect everything, and try to get full coverage on that, ensuring we get the behaviour we expect.